### PR TITLE
wrote getPairOrientation method for GATKRead

### LINKED ISSUE #1550: Add new functionality in GATK4 to directly extract SamPairUtil.PairOrientation from GATKRead
Currently such information can only be extracted indirectly via a conversion to SamRecord, which incurs unnecessary performance penalty:

```
GATKRead read = ... // proper initialization
SAMFileHeader header = ...
SamPairUtil.PairOrientation readOrientation = SamPairUtil.getPairOrientation(read.convertToSAMRecord(header))
```

### PR DESCRIPTION
**Pull request by davidbenjamin**
Closes #1550.  @SHuang-Broad do you still want this?

### CODE REVIEW

**Comment by SHuang-Broad** [comment]
sorry, David, just noticed this.

**Comment by davidbenjamin** [comment]
@SHuang-Broad `getFragmentLength()` is supposed to return the mate's 5' position minus the read's 5' position, so `getStart() + getFragmentLength()` should equal the mate end for a forward strand read.  Therefore, it seems to me that regardless of overlap the method is comparing 5' coordinates correctly, provided that we trust `getFragmentLength()`.

File: src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java [hunk 1]
```diff
@@ -1378,5 +1378,95 @@ public void testUnmappedUnpairedDistinctionmates(final GATKRead read, final bool
...
+        forwardStrandTandem.setPosition(new SimpleInterval("1", 100, 200));
+        forwardStrandTandem.setMatePosition(new SimpleInterval("1", 200, 300));
+
+        final GATKRead reverseStrandTandem = ArtificialReadUtils.createArtificialRead("100M");
+        reverseStrandTandem.setIsPaired(true);
+        reverseStrandTandem.setIsReverseStrand(true);
+        reverseStrandTandem.setMateIsReverseStrand(true);
+        reverseStrandTandem.setPosition(new SimpleInterval("1", 100, 200));
+        reverseStrandTandem.setMatePosition(new SimpleInterval("1", 200, 300));
+
```
Comment (SHuang-Broad) [inline src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: May I also suggest adding test cases:
* mates mapped to different chromosomes
Of course if you believe getting pair orientation from such mate pairs to be non-sense, then it should be move to the test cases expecting exceptions.

File: src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java [hunk 2]
```diff
@@ -1378,5 +1378,95 @@ public void testUnmappedUnpairedDistinctionmates(final GATKRead read, final bool
...
+    @DataProvider(name = "pair_orientation")
+    public Object[][] pairOrientation() {
+        final GATKRead forwardStrandInnie = ArtificialReadUtils.createArtificialRead("100M");
+        forwardStrandInnie.setIsPaired(true);
+        forwardStrandInnie.setIsReverseStrand(false);
+        forwardStrandInnie.setMateIsReverseStrand(true);
+        forwardStrandInnie.setPosition(new SimpleInterval("1", 100, 200));
+        forwardStrandInnie.setMatePosition(new SimpleInterval("1", 200, 300));
+        forwardStrandInnie.setFragmentLength(200);
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java [hunk 3]
```diff
@@ -600,6 +601,24 @@ default int numCigarElements(){
      */
     void setIsSupplementaryAlignment( final boolean isSupplementaryAlignment );
 
+    /**
+     * Computes the pair orientation
+     * @throws IllegalArgumentException If the read is not paired, or if either read or mate is unmapped
+     */
+    default SamPairUtil.PairOrientation getPairOrientation() {
+        final boolean readIsOnReverseStrand = isReverseStrand();
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java]: done

File: src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java (hunk 1 above)
Comment (davidbenjamin) [reply src/test/java/org/broadinstitute/hellbender/utils/read/GATKReadAdaptersUnitTest.java]: @SHuang-Broad What's your opinion?  It doesn't seem like any of the existing values of `SamPairUtil.PairOrientation` make sense.  If it weren't in htsjdk I would be tempted to add `DIFFERENT_CHROMOSOMES` to the enum.  I could imagine throwing an error and leaving it to the calling code to check.  I could also imagine having this method return an `Optional<SamPairUtil.PairOrientation>` and reserving an empty result for this case.

**Comment by davidbenjamin** [comment]
@SHuang-Broad Could you explain your disperse duplicated / improperly paired scenario?  It seems to me that if a read and its mate map to the same strand then they have to be oriented in the same (5-prime to 3-prime) direction.  

I quickly looked into the improperly paired flag and the word online is that there's no fixed definition.  Sometimes it just means an innie with larger-than-expected fragment size.

I will do whatever you want me to do, because I'm just trying to close old GATK issues!

**Comment by SHuang-Broad** [comment]
Yes, we need to talk and draw on a board when we are both in office. It's a bit hard to explain via pure text here.

**Comment by davidbenjamin** [comment]
Sounds good!

**Pull request merged**